appscale/tools/appscale_tools.py

```diff
@@ -654,3 +686,145 @@ def upload_app(cls, options):
+      cls.shut_down_appscale_if_running(options)
+      cls.run_upgrade_script(options, upgrade_version_available, master_ip, zk_ips, db_ips)
+      return
+
+    cls.shut_down_appscale_if_running(options)
+    cls.upgrade_appscale_code(options)
+    cls.run_upgrade_script(options, upgrade_version_available, master_ip, zk_ips, db_ips)
+
+  @classmethod
+  def run_upgrade_script(cls, options, upgrade_version_available, master_ip, zk_ips, db_ips):
+    """ Runs the upgrade script which checks for any upgrades needed to be performed.
+      Args:
+        options: A Namespace that has fields for each parameter that can be
+          passed in via the command-line interface.
+        upgrade_version_available: The latest version available to upgrade to.
+        master_ip: The IP address to the head node.
+        zk_ips: A list of ZooKeeper node IPs.
+        db_ips: A list of database node IPs.
+    """
+    ts = time.time()
+    timestamp = datetime.datetime.fromtimestamp(ts).strftime('%Y-%m-%d_%H:%M:%S')
+
+    zk_ips_str = cls.get_ip_str_for_command(zk_ips)
+    db_ips_str = cls.get_ip_str_for_command(db_ips)
+    master_ip_str = cls.get_ip_str_for_command(master_ip)
+    upgrade_script_command = cls.UPGRADE_SCRIPT + " " + upgrade_version_available + \
+      " " + options.keyname + " " + timestamp + " " + "--master" +  " " + master_ip_str + \
+      " " + "--zookeeper" + " " + zk_ips_str + \
+      " " + "--database" + " " + db_ips_str
+
+    AppScaleLogger.log("Running upgrade script to check if any other upgrade is needed.")
+    try:
+      RemoteHelper.ssh(master_ip, options.keyname, upgrade_script_command,
+        options.verbose)
+      upgrade_status_file = cls.UPGRADE_STATUS_FILE_LOC + timestamp + cls.JSON_FILE_EXTENTION
+      command = 'cat' + " " + upgrade_status_file
+      ssh_file = RemoteHelper.get_command_output_from_remote(master_ip, command, options.keyname)
+      upgrade_status = ssh_file.stdout.read()
+
+      json_status = json.loads(upgrade_status)
+      if not json_status.keys():
+        AppScaleLogger.log("AppScale is already at its latest version.")
+
+      for key in json_status.keys():
+        for second_level_key in json_status[key]:
+          if second_level_key == 'Completion Status' and \
+            json_status[key][second_level_key] == 'Success':
+            AppScaleLogger.success("{} process was successfully executed.".format(key))
+          elif not json_status[key][second_level_key] == 'Success':
+            AppScaleLogger.warn("Error encountered during upgrade process -> {} : {}".
+              format(second_level_key, json_status[key][second_level_key]))
+            AppScaleLogger.warn("For more information refer to " + upgrade_status_file
+              + " by logging into your head node.")
+
+    except ShellException:
+      AppScaleLogger.warn("Error executing upgrade script.")
+
+  @classmethod
+  def get_ip_str_for_command(cls, ip_list):
```

Comment:
The problem with using json.dumps(list) or converting the list to string using any existing utilities is that it leaves spaces between list items. And because I am sending the list as an argument to the python script, it interprets the spaces as args too. Hence this function which builds the list without spaces.